README.md
# Splice site annotation and prediction

## Background

There has been countless approaches at predicting the deleteriousness of a variant on splicing. MaxEntScan [1] for example can preform this prediction when the score of the wild type is compared to the variant type. As shown by Jian in 2014 [2] and Fontenelle in 2015 [3], this algorithm is not only the best, but also better in isolation than combined with others.

[1] G. Yeo and C. B. Burge, “**Maximum entropy modeling of short sequence motifs with applications to RNA splicing signals**,” *J. Comput. Biol.*, vol. 11, no. 2–3, pp. 377–94, Jan. 2004.

[2] X. Jian, E. Boerwinkle, and X. Liu, “**In silico prediction of splice-altering single nucleotide variants in the human genome**,” *Nucleic Acids Res.*, vol. 42, no. 22, pp. 13534–13544, Nov. 2014.

[3] H. Fontenelle, M. C. Eike, T. Håndstad, E. Skorve, S. T. Koksrud Seljebotn, T. Grünfeld and D. E. Undlien, "**Best practice for use of splice variant prediction tools**," *4th course in Next Generation Sequencing*, Bertinoro, Italy, 12-16 May 2015 [PDF] (awarded Best Poster)

This library automates the annotation of variants with MaxEntScan [1]. It predicts the effect by calculating the ratio between the wild score and variant type score. The site is predicted to be lost when that ratio is below -21.4% [2].

## Annotation

### Effect on authentic splice site
```
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO
3   30686236    id1 CA  AT  .   .   splice=NM_003242.5|predicted_lost|10.82|0.0
3    30686405    id2    ATG    TTC  .    .   splice=NM_003242.5|predicted_lost|9.26|5.42
3   30686258    id3 C   T   .   .   splice=NM_003242.5|no_effect|10.82|10.82
```

- `splice=` is the INFO keyword for this annotation, it is followed by the transcript for which this prediction is valid, the predicted effect, the MaxEntScan score for the wild type and the score for the variant type. The fields are separated by `|` (pipes).
- `predicted_lost` means the the variant score dropped significantly with respect to the wild type score.
- `predicted_conserved` means the score did not drop significantly, or not at all
- `consensus_not_affected` means the nucleotide sequence of the authentic splice site was not affected by the variant.
From Burge's paper for the MaxEntScan implementation, we know that "5' donors" are 9-mer with a consensus GT after the junction; and that "3' acceptors" are 23-mer with a consensus AG before the junction. Sites that do not conform to the canonical AG/GT are not taken into consideration (there exists no known algorithm that can).
- SNP's within +/- 2bp from junctions are annotated with `splice_donor_variant` or `splice_acceptor_variant` in a way similar to VEP.
- In addition, a variant may be annotated with `not_transcribed` if it falls outside a transcript; or with `NA` if something has failed.

### De novo: putative new splice Sites

```
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO
2   162060116    id4 T  G  .   .  splice=NM_004180.2|no_effect|9.79|9.79&NM_004180.2|de_novo|0.0|4.7|9.79|10
2    162041853    id5    T    G .    .   splice=NM_004180.2|no_effect|6.64|6.64&NM_004180.2|de_novo|0.0|10.86|6.64|5581
```

Several effects are joined with the `&` character. Here the variant are scored for their effect on the authentic splice site, but in addition "de novo" sites were found. Following the `de_novo` keyword is the transcript name and the MaxEntScan score of the wild sequence in that position, the score of the de novo site in that position, the score of the closest authentic site, and the distance (in nucleotides) to the closest authentic site.
The rules for marking a de novo are:
* If MES score for REF is 0, report "de novo" if score for ALT is ≥4
* If MES score for REF is >0 , report "de novo" if score for ALT is ≥ 25% increased compared to REF score
The score of the authentic closest site is not taken into consideration in those rules, but it is usually considered important [4] so we added it here.

[4] C. Houdayer, “**In Silico Prediction of Splice-Affecting Nucleotide Variants**,” in *In Silico Tools for Gene Discovery*, vol. 760, B. Yu and M. Hinchcliffe, Eds. Totowa, NJ: Humana Press, 2011, pp. 269–281.

Notes:
* De novo searches for Donors if the closest authentic site is a Donor. Conversely it searches for acceptors if the closest is an Acceptor.
* Only AG/GT are searched since the score of any other sequence is meaningless.
* Multi-allelic sites are scored/predicted as well, and separated by `,` (comma)

## Command-line tools

```
$ python splice_annotate.py -h
usage: splice_annotate.py [-h] -i INPUT [-o OUTPUT] [--genepanel GENEPANEL]
                         [--refseqgene REFSEQGENE] [--refseq REFSEQ]

Annotate a VCF with splice site effect prediction.

optional arguments:
 -h, --help            show this help message and exit
 -i INPUT, --input INPUT
                       Input VCF file
 -o OUTPUT, --output OUTPUT
                       Output VCF file for annotation of predictions
                       (default=same as input)
 --genepanel GENEPANEL
                       Filepath for genepanel
 --refseqgene REFSEQGENE
                       Filepath for refSeqGene
 --refseq REFSEQ       Filepath for refSeq reference FASTA sequences
```


CHANGES.txt
# Release notes

## v0.3, 09-Jun-16 -- version 0.3

### Bug fixes:

- header is incomplete for de_novo
- transcript missing for splice_donor_variant
- missing splice_acceptor_variant
- fix README

## v0.2, 22-Feb-16 -- version 0.2

## v0.1, 13-Apr-15 -- version 0.1


splice/refseq_utils.py
# -*- coding: utf-8 -*-
"""
Created on Wed Dec  3 12:39:25 2014

@author: huguesfo
"""


import os
import csv

from pyfaidx import Fasta

email = "hugues.fontenelle@medisin.uio.no"

# Human Genome Assembly Data
# http://www.ncbi.nlm.nih.gov/projects/genome/assembly/grc/human/data/
RefSeqAccession = {'hg19': {
    '1': "NC_000001.10",    '2': "NC_000002.11",    '3': "NC_000003.11",
    '4': "NC_000004.11",    '5': "NC_000005.9",    '6': "NC_000006.11",
    '7': "NC_000007.13",    '8': "NC_000008.10",    '9': "NC_000009.11",
    '10': "NC_000010.10",    '11': "NC_000011.9",    '12': "NC_000012.11",
    '13': "NC_000013.10",    '14': "NC_000014.8",    '15': "NC_000015.9",
    '16': "NC_000016.9",    '17': "NC_000017.10",    '18': "NC_000018.9",
    '19': "NC_000019.9",    '20': "NC_000020.10",    '21': "NC_000021.8",
    '22': "NC_000022.10",    'X': "NC_000023.10",    'Y': "NC_000024.9",
},
'hg38': {
    '1': "NC_000001.11",    '2': "NC_000002.12",    '3': "NC_000003.12",
    '4': "NC_000004.12",    '5': "NC_000005.10",    '6': "NC_000006.12",
    '7': "NC_000007.14",    '8': "NC_000008.11",    '9': "NC_000009.12",
    '10': "NC_000010.11",    '11': "NC_000011.10",    '12': "NC_000012.12",
    '13': "NC_000013.11",    '14': "NC_000014.9",    '15': "NC_000015.10",
    '16': "NC_000016.10",    '17': "NC_000017.11",    '18': "NC_000018.10",
    '19': "NC_000019.10",    '20': "NC_000020.11",    '21': "NC_000021.9",
    '22': "NC_000022.11",    'X': "NC_000023.11",    'Y': "NC_000024.10",
}
}

try:
    import cPickle as pickle
except:
    import pickle

__all__ = ["memoize"]

def memoize(function, limit=None):
    if isinstance(function, int):
        def memoize_wrapper(f):
            return memoize(f, function)

        return memoize_wrapper

    dict = {}
    list = []
    def memoize_wrapper(*args, **kwargs):
        key = pickle.dumps((args, kwargs))
        try:
            list.append(list.pop(list.index(key)))
        except ValueError:
            dict[key] = function(*args, **kwargs)
            list.append(key)
            if limit is not None and len(list) > limit:
                del dict[list.pop(0)]

        return dict[key]

    memoize_wrapper._memoize_dict = dict
    memoize_wrapper._memoize_list = list
    memoize_wrapper._memoize_limit = limit
    memoize_wrapper._memoize_origfunc = function
    memoize_wrapper.func_name = function.func_name
    return memoize_wrapper


# ------------------------------------------------------------
def chr_to_refseq(chrom, ref='hg19'):
    '''Returns the NCBI Reference SequenceID from chromosome number'''
    return RefSeqAccession[ref].get(chrom, "Unkwnown chromosome number")


# ------------------------------------------------------------
def get_fasta(chrom, start, end, refseq=None, ref='hg19'):
    '''Retrieves FASTA sequence'''
    fasta = ''
    chrom = str(chrom)
    if chrom.startswith("chr"): chrom = chrom[3:]
    if refseq:
        '''Retrieves FASTA sequence from local refSeq file'''
        if os.path.isfile(refseq):
            genome = Fasta(refseq)
            fasta = str(genome[chrom][start-1:end])
    else:
        '''Retrieves FASTA sequence from NCBI (internet connection)'''

        from Bio import Entrez, SeqIO

        refseq_id = chr_to_refseq(chrom, ref)
        Entrez.email = email
        handle = Entrez.efetch(db="nucleotide",
                               id=refseq_id,
                               rettype="fasta",
                               strand=1,
                               seq_start=start,
                               seq_stop=end)
        entrez_record = SeqIO.read(handle, "fasta")
        handle.close()
        fasta = str(entrez_record.seq)

    return fasta


# ------------------------------------------------------------
def get_auth_from_refseqgene(chrom, pos, refseqgene=None,  ref='hg19'):
    '''Retrieves exon definitions from local refSeqGene file'''
    refSeq = []
    try:
        with open(refseqgene, 'r') as f:
            refSeqList = csv.reader(f, delimiter='\t')
            for row in refSeqList:
                if row[0].strip().startswith('#'):
                    continue
                if row[2][3:] == chrom and int(row[4]) <= pos <= int(row[5]):
                    refSeq = row
                    break
    except Exception as ex:
        raise ex

    if refSeq == []:
        print 'chr%s:%s Not in a RefSeq transcript\n' % (chrom, str(pos))
        return None

    exonList = [(int(x), int(y)) for (x,y) in zip(refSeq[9].rstrip(',').split(','), refSeq[10].rstrip(',').split(','))]
    strand = refSeq[3]
    transcript = refSeq[1]

    rel_exons = [(abs(start-pos), abs(end-pos)) for (start, end) in exonList]
    a = [(min(e), e.index(min(e))) for i, e in enumerate(rel_exons)]
    id0 = a.index(min(a))
    id1 = min(a)[1]
    closest_splice_site = exonList[id0][id1]

    if strand == '+':
        splice_type = {0: "Acceptor", 1: "Donor"}[id1]
    else:  # the convention is opposite for - strand!!
        splice_type = {1: "Acceptor", 0: "Donor"}[id1]

    return {'chrom': chrom,
            'pos': closest_splice_site, 'splice_type': splice_type,
            'strand': strand, 'transcript': transcript}


# ------------------------------------------------------------
def get_auth_from_genepanel(chrom, pos, genepanel=None, ref='hg19'):
    '''Retrieves exon definitions from local genepanel file'''
    gp = []
    try:
        with open(genepanel, 'r') as f:
            gpList = csv.reader(f, delimiter='\t')
            for row in gpList:
                if row[0].strip().startswith('#'):
                    continue
                txStart, txEnd = int(row[1]), int(row[2])
                if row[0] == chrom and txStart <= pos <= txEnd:
                    gp = row
                    break
    except Exception as ex:
        raise ex

    if gp == []:
        print 'chr%s:%s Not in a genepanel transcript\n' % (chrom, str(pos))
        return None

    exonList = [(int(x), int(y)) for (x,y) in zip(gp[12].rstrip(',').split(','), gp[13].rstrip(',').split(','))]
    strand = gp[5]
    transcript = gp[3]

    rel_exons = [(abs(start-pos), abs(end-pos)) for (start, end) in exonList]
    a = [(min(e), e.index(min(e))) for i, e in enumerate(rel_exons)]
    id0 = a.index(min(a))
    id1 = min(a)[1]
    closest_splice_site = exonList[id0][id1]

    if strand == '+':
        splice_type = {0: "Acceptor", 1: "Donor"}[id1]
    else:  # the convention is opposite for - strand!!
        splice_type = {1: "Acceptor", 0: "Donor"}[id1]

    return {'chrom': chrom, 'pos':closest_splice_site, 'splice_type': splice_type,
            'strand': strand, 'transcript': transcript}


# ------------------------------------------------------------
def get_auth_from_NCBI(chrom, pos, ref='hg19'):
    '''Retrieves exon definitions from from NCBI (internet connection)'''
    from Bio import Entrez, SeqIO

    refseq_id = chr_to_refseq(chrom, ref)
    startpos = pos - 5000
    endpos = pos + 5000
    Entrez.email = email
    handle = Entrez.efetch(db="nucleotide",
                           id=refseq_id,
                           rettype="gb",
                           retmode="text",
                           seq_start=startpos,
                           seq_stop=endpos)
    entrez_record = SeqIO.read(handle, "gb")
    handle.close()

    # find mRNA exons start:end in the subsequence
    # exons = {}
    exonList = []
    for feature in entrez_record.features:
        if feature.type == 'mRNA':
            if not feature.sub_features:
                exonList.append((feature.location.start.position + startpos,
                                 feature.location.end.position + startpos,
                                 feature.location.strand))
            else:
                for sub_feature in feature.sub_features:
                    exonList.append((sub_feature.location.start.position + startpos,
                                     sub_feature.location.end.position + startpos,
                                     sub_feature.location.strand))
    if not exonList:
        print 'chr%s:%s No authentic splice site nearby\n' % (chrom, str(pos))
        return None

    rel_exons = [(abs(start-pos), abs(end-pos)) for (start, end, s) in exonList]
    a = [(min(e), e.index(min(e))) for i, e in enumerate(rel_exons)]
    id0 = a.index(min(a))
    id1 = min(a)[1]
    closest_splice_site = exonList[id0][id1]

    strand = {1: "+", -1: "-"}[exonList[id0][2]]
    if strand == '+':
        splice_type = {0: "Acceptor", 1: "Donor"}[id1]
    else:  # the convention is opposite for - strand!!
        splice_type = {1: "Acceptor", 0: "Donor"}[id1]

    return {'chrom': chrom, 'pos': closest_splice_site-1,
            'splice_type': splice_type,
            'strand': strand, 'transcript': refseq_id}


# ------------------------------------------------------------
@memoize(10)
def get_closest_authentic(chrom, pos, refseqgene=None, genepanel=None, ref='hg19', get_sequence=False, refseq=None, seq_size=50):
    '''
    Find the closest intron-exon junction in the RefSeq w.r.t. the mutation
    position. Returns a dictionnary with the splice site position, splice type,
    strand for a transcript.
    Eg:
    {'pos': 32944538, 'splice_type': 'Acceptor', 'strand': '+', 'transcript': 'NM_000059'}
    '''
    if chrom.startswith("chr"): chrom = chrom[3:]
    if refseqgene:
        auth = get_auth_from_refseqgene(chrom, pos, refseqgene=refseqgene, ref='hg19')
    elif genepanel:
        auth = get_auth_from_genepanel(chrom, pos, genepanel=genepanel, ref='hg19')
    else:
        auth = get_auth_from_NCBI(chrom, pos, ref='hg19')

    if auth and get_sequence:
        auth['fasta'] = get_fasta(chrom=auth['chrom'], start=auth['pos']-seq_size, end=auth['pos']+seq_size, refseq=refseq)

    return auth


# ============================================================
if __name__ == "__main__":
    pass
